src/main/kotlin/com/openlattice/graph/partioning/RepartitioningJob.kt

```diff
@@ -0,0 +1,352 @@
+import java.sql.PreparedStatement
+import java.util.*
+
+/**
+ * Background job for re-partitioning data in an entity set.
+ *
+ * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
+ */
+class RepartitioningJob(
+        state: RepartitioningJobState
+) : AbstractDistributedJob<Long, RepartitioningJobState>(state), MetastoreAware {
+    @JsonCreator
+    constructor(
+            id: UUID?,
+            taskId: Long?,
+            status: JobStatus,
+            progress: Byte,
+            hasWorkRemaining: Boolean,
+            result: Long?,
+            state: RepartitioningJobState
+    ) : this(state) {
+        initialize(id, taskId, status, progress, hasWorkRemaining, result)
+    }
+
+    constructor(
+            entitySetId: UUID,
+            oldPartitions: List<Int>,
+            newPartitions: Set<Int>
+    ) : this(RepartitioningJobState(entitySetId, oldPartitions, newPartitions))
+
+    private var phase = 0
+
+    @Transient
+    private lateinit var hds: HikariDataSource
+
+    @Transient
+    private lateinit var entitySets: IMap<UUID, EntitySet>
+
+    private val currentlyMigratingPartition: Int
+        get() = state.oldPartitions[state.currentlyMigratingPartitionIndex]
+
+    @JsonIgnore
+    override fun setHazelcastInstance(hazelcastInstance: HazelcastInstance) {
+        super.setHazelcastInstance(hazelcastInstance)
+        this.entitySets = HazelcastMap.ENTITY_SETS.getMap(hazelcastInstance)
+    }
+
+    @JsonIgnore
+    override fun setHikariDataSource(hds: HikariDataSource) {
+        this.hds = hds
+    }
+
+    override fun initialize() {
+        state.needsMigrationCount = getNeedsMigrationCount()
+    }
+
+    override fun processNextBatch() {
+        if (state.needsMigrationCount == 0L) {
+            hasWorkRemaining = false
+            return
+        }
+
+        /**
+         * Do an INSERT INTO ... SELECT FROM to re-partition the data.
+         *
+         * Entity key id depenedent operations will not see data, until data has been inserted to the appropriate partition.
+         */
+
+        state.repartitionCount += repartition(REPARTITION_DATA_SQL)
+        state.repartitionCount += repartition(REPARTITION_IDS_SQL)
+        state.repartitionCount += repartition(REPARTITION_EDGES_SQL)
+
+        /**
+         * Phase 1
+         * Delete data whose partition doesn't match it's computed partition.
+         */
+        if (phase == 1) {
+            state.deleteCount += delete(DELETE_DATA_SQL)
+            state.deleteCount += delete(DELETE_IDS_SQL)
+            state.deleteCount += delete(DELETE_EDGES_SQL)
+        }
+
+        result = state.repartitionCount + state.deleteCount
+        hasWorkRemaining = (++state.currentlyMigratingPartitionIndex < state.oldPartitions.size)
+
+        //TODO: Consider adding completion hook to distributable jobs framework
+        //Once we are done, set the partitions.
+        if (!hasWorkRemaining && phase == 0) {
```

Comment:
Does phase 0 mean complete? (It might be helpful to use en enum or constants to name these phases)

Unrelatedly, what's the reason for waiting until completion to update the partitions on the entity set itself? Won't any incoming writes during the repartition be competing with the repartition job? (I realize we control these jobs so we could just shut down integrations while running, but seems like if there's a way to minimize/reduce interruption, it would be preferable)

Reply:
It's to maximize availability. If we set partitions, then getEntities call will use newPartitions and reads will fail. This way most of the data is copied and there is instead a small window where stale reads may happen.